parsl/curvezmq.py

```diff
@@ -136,6 +133,12 @@ def __init__(self, cert_dir: Optional[Union[str, os.PathLike]]) -> None:
         if self.encrypted:
             self.auth_thread = self._start_auth_thread()
 
+    def __del__(self):
+        # Avoid issues in which the auth_thread attr was
+        # previously deleted
+        if getattr(self, "auth_thread", None):
+            self.auth_thread.stop()
+
```

Comment:
This isn't on the base class; perhaps just:
```python
if self.auth_thread:
    self.auth_thread.stop()
```
On the other hand, I don't see the need for this method at all.  If it's just for cleanup purposes, then the thread is already stopped when this object is finalized.  If anything, if we want to manually stop the thread (which I don't have a problem with in theory), then I'd rather we implement a proper interface for these classes, such as a `.shutdown()` method, or `__enter__()`/`__exit__()` methods.

Reply:
The use of `getattr` is an extra precaution to avoid warnings ([destructor exceptions are ignored](https://docs.python.org/3/reference/datamodel.html#object.__del__)) in the unlikely event that the `auth_thread` attribute gets deleted.

The method is not strictly necessary, but the underlying `zmq.Context` object has it's own `__del__` method, which closes sockets via `self.destroy()`. We need stop the auth thread before attempting to close sockets.